### PR TITLE
Drop .NET Standard 1.3 and rework AsyncEvents

### PR DESCRIPTION
**Pull request by Emzi0767**
# Summary
This PR drops support for .NET Standard 1.3, per previous poll. In addition to it, it also reworks AsyncEvents.

# Details
.NET Standard 1.3 target is no more. The library targets .NET Standard 2.0 exclusively now. This means that .NETFX version requirements are >=4.6.1 now (as opposed to 4.6 previously). Do note that FX is not explicitly supported; you should be using .NET Core.

The `AsyncEvent` implementation in the library has been replaced with one from [Emzi0767.Common](https://www.nuget.org/packages/Emzi0767.Common/). The new implementation is far more flexible, robust, and comes with more features. Event handlers all now take 2 arguments, instead of 1: a `sender` argument was added as a first argument (the second argument is the prior sole argument). With this change, comes removal of `Client` from most `EventArgs` instances. The instance is now provided, either directly or indirectly, via `sender` argument of event handlers now.

# Changes proposed
- Drop .NET Standard 1.3 target.
   - This bumps .NETFX req to 4.6.1.
   - This allows us to evolve better.
- Use `AsyncEvent` from `Emzi0767.Common`.
   - Event handlers now take 2 arguments: `sender` and `args`.
      - First one is the sending instance (typically client or the extension which generated the event).
      - Second one is the arguments to the event.
   - `Client` property is now removed from most `EventArgs` instances. See above on how to get the client instance.
   - Events now have a timeout. The timeout is soft, when it expires, an error log will be emitted to the configured logger.
   - Event errors are more detailed, exceptions are unwrapped, and the method which caused the exception is now listed in the error log message.

# Notes
Yes I am farming.

### CODE REVIEW

**Comment by Emzi0767** [comment]
Build fails because AppVeyor doesn't know how to pull a PR properly.

File: DSharpPlus/Net/WebSocket/WebSocketClient.cs [hunk 1]
```diff
@@ -291,50 +292,46 @@ public static IWebSocketClient CreateNew(IWebProxy proxy)
...
-        private void EventErrorHandler(string evname, Exception ex)
-        {
-            if (evname.ToLowerInvariant() == "ws_error")
-                Console.WriteLine($"WSERROR: {ex.GetType()} in {evname}!");
-            else
-                this._exceptionThrown.InvokeAsync(new SocketErrorEventArgs(null) { Exception = ex }).ConfigureAwait(false).GetAwaiter().GetResult();
-        }
+        private void EventErrorHandler<TArgs>(AsyncEvent<WebSocketClient, TArgs> asyncEvent, Exception ex, AsyncEventHandler<WebSocketClient, TArgs> handler, WebSocketClient sender, TArgs eventArgs)
+            where TArgs : AsyncEventArgs
+            => this._exceptionThrown.InvokeAsync(this, new SocketErrorEventArgs() { Exception = ex }).ConfigureAwait(false).GetAwaiter().GetResult();
```
Comment (Neuheit) [inline DSharpPlus/Net/WebSocket/WebSocketClient.cs:334]: Why does error handling need to be handled synchronously compared to the other event invokers?

File: DSharpPlus/Clients/DiscordClient.Events.cs [hunk 2]
```diff
@@ -1,72 +1,79 @@
...
-using Microsoft.Extensions.Logging;
 using DSharpPlus.EventArgs;
+using Emzi0767.Utilities;
+using Microsoft.Extensions.Logging;
 
 namespace DSharpPlus
 {
     public sealed partial class DiscordClient
     {
+        internal static TimeSpan EventExecutionLimit { get; } = TimeSpan.FromSeconds(1);
```
Comment (Neuheit) [inline DSharpPlus/Clients/DiscordClient.Events.cs:10]: I feel like there should be an option for this in the DiscordConfiguration.

Comment (Emzi0767) [reply DSharpPlus/Clients/DiscordClient.Events.cs:10]: It won't be. The limit is soft, and it's there to inform the user they are stalling the queue.

File: DSharpPlus/Net/WebSocket/WebSocketClient.cs (hunk 1 above)
Comment (Emzi0767) [reply DSharpPlus/Net/WebSocket/WebSocketClient.cs:334]: The upstream intent here was logging etc. The previous implementation did the same.

**Pull request merged**